### ISSUE TITLE
[Feature Request] Audio Stream Interrupt/Commandeer

### ISSUE THREAD

**Issue by klavavaski** [markokajzer/discord-soundbot#80]
This.... this is kind of a weird one, and/or a lot of work to implement, in two stages. If you close this outright as wholly unnecessary functionality, I'll totally understand.

Stage 1: A command to interrupt the current playback and immediately start with another sound. This could probably just be a new command that combines the functionality of !stop with a playback argument. I might just hack one together later.
Stage 1 Optional: Maybe this also only interrupts the queue at the current sound, and then continues the rest of the playback stack. Haven't looked at how command parsing is handled, so maybe it's less a "queue" and more the bot only processing one command at a time, but I'm thinking it's probably a queueing mechanism, since !stop will proc regardless of the current chat state, and there is a !skip/!next
Stage 2: The same as stage 1, but starting playback of the next sound at the same timestamp as the interrupted stream, or a specified timestamp. This one is tricky for me, because I haven't taken the time to actually look at how playback happens. Is this getting piped into a fifo stream? If so, it may be possible to hijack the fifo stream and inject a different audio file. This will almost definitely end up with audio artifacting, because fifos are fun that way

**Comment by markokajzer** [comment]
Hey man, thanks for the suggestions. I think Stage 1 is actually already covered with the `!next` command. You can do `!next mySound` to interrupt the current one with a new sound that'll immediately start playing. After that the queue will continue as normal.

For the second one... That sounds like a pretty weird use case to me and honestly not like something that I would like to implement. As you said, it'd most likely come with artifacts which would probably damper the experience.

Let me know if `!next` covers your use cases.

**Comment by klavavaski** [comment]
definitely caught stage 1 with !next, which i didn't even realize the functionality of. love that it keeps the queue intact.

re: stage2, not sure if you ever dealt with the mumble soundboard bot from erulabs, but it handled interrupts significantly less gracefully.

basically, the fifo pipe opens and plays, and on typing a second !play equivalent, the contents of a different audio stream would get injected in place of the original. absolutely insane combo mechanics. i might look into figuring out a way to do something like...

interrupt playback with last half of specified sound clip:
`!interrupt 10secondclip 00:00:05.000`

combo interrupt, specifying first file, timestamp to interrupt, second file, optional timestamp to start second file
`!comborupt firstfile 00:00:03.000 secondfile 00:00:06.000`

we should leave this closed since the functionality is super niche, but i'll let you know if i figure anything out.

**Comment by markokajzer** [comment]
To be honest, that sounds to me like an incredibly bad interface.

Who in their right mind would **regularly** type something incredibly tedious like `!comborupt firstfile 00:00:03.000 secondfile 00:00:06.000`? If one would want to use this specific configuration why not either a) create an additional sound that combos the two together as specified, or b) create shorter clips and tie them together with the `!combo shortfirstfile shortsecondfile`. This is in my mind much more comfortable to use than specifying exact timestamps on the fly, even with creating one additional sound. Otherwise you would always have to look up the time again, or save `!comborupt` commands in additional textfiles.

Just because we can do something, does not mean we should. My 2 cents.

**Comment by klavavaski** [comment]
finally got a chance to come back to this, and you've got me rolling with laughter.

> Who in their right mind would **regularly** type something incredibly tedious

absolutely true. it comes more into play with the idea of our community soundboard having grown to over 3000 sound files, allowing for some really unique combo ideas. the more incredible interruptions have already been spliced as individual sound files.

...coincidentally, some of the members of that community did curate text files with multiple lines of combos that they found interesting/funny, which circles right back to: "who in their right mind?"

admittedly, the example command sets i gave are... fundamentally clunky. writing out those longform timestamps every single time you want to `!comborupt` is questionable already, let alone the fact that i left the hour and minute markers in the timestamps (i can't think of a good reason to have any file exceeding 15 seconds, and even that's pushing it)

i suppose a more user-friendly version would only use seconds and fractions of seconds, but `!comborupt firstfile 3.000 secondfile 6.300` isn't much better, and still suffers from the same stumbles you already mentioned

maybe i'll introduce it with optional variables at the tail. something like:
`!comborupt firstfile secondfile` play firstfile for 3 seconds, interrupt with secondfile at the 3 second marker
`!comborupt firstfile secondfile 4.500` same as the first, but interrupt at the 4.5 second timestamp of firstfile
`!comborupt firstfile secondfile 4.500 6.500` adding secondfile starting timestamp to the previous example

or maybe an addendum to `!next`, an optional flag for where to start the playback for the second file:
`!firstfile`
`!next secondfile 4.500`

also, very probable to have some fun interactions with `random`. when i get a chance to actually get stuck into it, i'll let you know if i come up with anything interesting. even if it's niche, there may be bits and pieces that are **actually** useful

TL;DR: you're 100% correct, but i'm probably going to do it anyways lmao

**Comment by markokajzer** [comment]
Thanks for your well written out answer and for the passion for this project. 🙇
